File: 01-Computer-Science/code/python/seb_cs/utf8.py
from __future__ import annotations


class Utf8Error(ValueError):
    def __init__(self, message: str, offset: int) -> None:
        super().__init__(f"{message} at {offset}")
        self.offset = offset


def encode_utf8(text: str) -> bytes:
    return text.encode("utf-8")


def decode_utf8(data: bytes) -> str:
    i = 0
    out: list[str] = []
    while i < len(data):
        b0 = data[i]
        if b0 <= 0x7F:
            out.append(chr(b0))
            i += 1
            continue
        if (b0 & 0xE0) == 0xC0:
            expected = 1
            code_point = b0 & 0x1F
            if b0 < 0xC2:
                raise Utf8Error("overlong 2-byte sequence", i)
        elif (b0 & 0xF0) == 0xE0:
            expected = 2
            code_point = b0 & 0x0F
        elif (b0 & 0xF8) == 0xF0:
            expected = 3
            code_point = b0 & 0x07
            if b0 > 0xF4:
                raise Utf8Error("code point above U+10FFFF", i)
        else:
            raise Utf8Error("invalid leading byte", i)
        if i + expected >= len(data):
            raise Utf8Error("truncated sequence", i)
        for j in range(1, expected + 1):
            bj = data[i + j]
            if (bj & 0xC0) != 0x80:
                raise Utf8Error("invalid continuation byte", i + j)
            code_point = (code_point << 6) | (bj & 0x3F)
        if expected == 2:
            if code_point < 0x800:
                raise Utf8Error("overlong 3-byte sequence", i)
            if 0xD800 <= code_point <= 0xDFFF:
                raise Utf8Error("surrogate code point", i)
        if expected == 3:
            if code_point < 0x10000:
                raise Utf8Error("overlong 4-byte sequence", i)
            if code_point > 0x10FFFF:
                raise Utf8Error("code point above U+10FFFF", i)
        out.append(chr(code_point))
        i += expected + 1
    return "".join(out)

File: 01-Computer-Science/code/typescript/src/utf8.ts
/** Strict UTF-8 encode/decode with explicit malformed-sequence errors. */

export class Utf8Error extends Error {
  constructor(
    message: string,
    readonly offset: number,
  ) {
    super(message);
    this.name = "Utf8Error";
  }
}

export function encodeUtf8(text: string): Uint8Array {
  return new TextEncoder().encode(text);
}

export function decodeUtf8(bytes: Uint8Array): string {
  let i = 0;
  let out = "";
  while (i < bytes.length) {
    const b0 = bytes[i]!;
    if (b0 <= 0x7f) {
      out += String.fromCharCode(b0);
      i += 1;
      continue;
    }
    let expected = 0;
    let codePoint = 0;
    if ((b0 & 0xe0) === 0xc0) {
      expected = 1;
      codePoint = b0 & 0x1f;
      if (b0 < 0xc2) throw new Utf8Error("overlong 2-byte sequence", i);
    } else if ((b0 & 0xf0) === 0xe0) {
      expected = 2;
      codePoint = b0 & 0x0f;
    } else if ((b0 & 0xf8) === 0xf0) {
      expected = 3;
      codePoint = b0 & 0x07;
      if (b0 > 0xf4) throw new Utf8Error("code point above U+10FFFF", i);
    } else {
      throw new Utf8Error("invalid leading byte", i);
    }
    if (i + expected >= bytes.length) throw new Utf8Error("truncated sequence", i);
    for (let j = 1; j <= expected; j++) {
      const bj = bytes[i + j]!;
      if ((bj & 0xc0) !== 0x80) throw new Utf8Error("invalid continuation byte", i + j);
      codePoint = (codePoint << 6) | (bj & 0x3f);
    }
    if (expected === 2) {
      if (codePoint < 0x800) throw new Utf8Error("overlong 3-byte sequence", i);
      if (codePoint >= 0xd800 && codePoint <= 0xdfff) {
        throw new Utf8Error("surrogate code point", i);
      }
    }
    if (expected === 3) {
      if (codePoint < 0x10000) throw new Utf8Error("overlong 4-byte sequence", i);
      if (codePoint > 0x10ffff) throw new Utf8Error("code point above U+10FFFF", i);
    }
    out += String.fromCodePoint(codePoint);
    i += expected + 1;
  }
  return out;
}
